plugin/completion.py

```diff
@@ -25,8 +25,8 @@ def run(self, edit: sublime.Edit, index: int, event: Optional[dict] = None) -> N
         if not detail or not documentation:
             # To make sure that the detail or documentation fields doesn't exist we need to resove the completion item.
             # If those fields appear after the item is resolved we show them in the popup.
-            session = self.best_session('completionProvider.resolveProvider')
-            if session:
+            session = self.session_by_name(session_name)
+            if session and session.has_capability('completionProvider.resolveProvider'):
```

Comment:
Hmm, this doesn't work with dynamic registration and a document selector. `session.has_capability` would return False while it could have been registered for a particular language id. The previous code:

```
self.best_session('completionProvider.resolveProvider')
```

took care of going through the listener.

You should probably modify `session_by_name` to take an extra optional parameter `capability`, look up the listener in `session_by_name`, then iterate over the session buffers filtering by the capability.

Reply:
I should probably add some tests for this document selector + dynamic registration thing to guard against this.

Reply:
I will try to wrap my head around this tomorrow 🙂

Reply:
I'll try to write an architecture.md file to explain the high-level ideas if I can find the time

Reply:
> You should probably modify session_by_name to take an extra optional parameter capability, look up the listener in session_by_name, then iterate over the session buffers filtering by the capability.

I don't see a need to filter out session by capability and here is why :)

1. We have two sessions: A, B
2. Both sessions have the `completionProvider.resolveProvider` capability.

When we trigger AC 
We have:
- 2 completion items from session A 
- 1 from session B.

```
completinItem1  (session A)
completinItem2  (session A)
completinItem3  (session B)
```

To resolve the completion item, we need to send the resolve request to a the session that Completion Item bellogs to.
For example you cannot rersolve completinItem1 with session B (because it was provided by Session A)

so I added the session_name when requesting documentaion popup to know to what session to send the resove compeltion request:
```
st_details += make_command_link("lsp_resolve_docs", "More", {"index": index, "session_name": session_name})
																			 ^^^^^^^^^^^^^^
```

Given what I just said,
my question to you is why do I need to add the filtering by capability, because I already know to what exact session name to send the request to?

Reply:
> Both sessions have the completionProvider.resolveProvider capability.

If both A and B use `client/registerCapability` for registering their `completionProvider` with a document selector, then

```
session.has_capability('completionProvider.resolveProvider')
```

will return false. `listener.has_capability_async` will return true.

Reply:
I tried a few times to get this done, but I got stuck. (I am not sure of what to change although you explained it 2 times)

Can this be done with a different PR?

Reply:
I'll look at it maybe tomorrow

Reply:
Servers like lua-language-server and OmniSharp register their completionProvider dynamically, so showing docs would break for those I guess.

Reply:
Raoul is there any way to modify the codebase so `session.has_capability('completionProvider.resolveProvider')` is also affected when we register capabilities dynamically.  
```
lus_session = self.session_by_name("lua-language-server")
lus_session.has_capability('completionProvider.resolveProvider') # returns False
# in the mean time the server did registered that capability dynamically,
# so the next time we call:
lus_session.has_capability('completionProvider.resolveProvider') # it returns True
```

Reply:
That doesn't work, because if an LS has two different document selectors, it might register completionProvider.resolveProvider for document selector A, but not for document selector B.

For example, SourceKit could register a colorProvider for swift files, but not for C/C++/Obj-C/Obj-C++ files. So an extra argument `view` is needed. But in any case, I think it's better to go through the listener for this anyway. See my latest commit.